File: apps/api/laboratorio-go/internal/dto/fornecedor_dto.go
package dto

import "time"

type CreateFornecedorRequest struct {
	Nome       string  `json:"nome" validate:"required"`
	Email      string  `json:"email" validate:"required,email"`
	Contato    *string `json:"contato"`
	TipoPessoa string  `json:"tipoPessoa" validate:"required,oneof=FISICA JURIDICA"`
	Documento  string  `json:"documento" validate:"required"`
	Status     string  `json:"status" validate:"omitempty,oneof=ATIVO INATIVO"`
	CEP        *string `json:"cep"`
	UF         *string `json:"uf"`
	Cidade     *string `json:"cidade"`
	Logradouro *string `json:"logradouro"`
	Numero     *string `json:"numero"`
	Bairro     *string `json:"bairro"`
	Observacao *string `json:"observacao"`
	UsuarioID  uint    `json:"usuarioId" validate:"required"`
}

type UpdateFornecedorRequest struct {
	Nome       string  `json:"nome" validate:"required"`
	Email      string  `json:"email" validate:"required,email"`
	Contato    *string `json:"contato"`
	TipoPessoa string  `json:"tipoPessoa" validate:"required,oneof=FISICA JURIDICA"`
	Documento  string  `json:"documento" validate:"required"`
	Status     string  `json:"status" validate:"omitempty,oneof=ATIVO INATIVO"`
	CEP        *string `json:"cep"`
	UF         *string `json:"uf"`
	Cidade     *string `json:"cidade"`
	Logradouro *string `json:"logradouro"`
	Numero     *string `json:"numero"`
	Bairro     *string `json:"bairro"`
	Observacao *string `json:"observacao"`
	UsuarioID  uint    `json:"usuarioId" validate:"required"`
}

type FornecedorResponse struct {
	ID         uint       `json:"id"`
	Nome       string     `json:"nome"`
	Email      string     `json:"email"`
	Contato    *string    `json:"contato"`
	TipoPessoa string     `json:"tipoPessoa"`
	Documento  string     `json:"documento"`
	Status     string     `json:"status"`
	CEP        *string    `json:"cep"`
	UF         *string    `json:"uf"`
	Cidade     *string    `json:"cidade"`
	Logradouro *string    `json:"logradouro"`
	Numero     *string    `json:"numero"`
	Bairro     *string    `json:"bairro"`
	Observacao *string    `json:"observacao"`
	CreatedAt  time.Time  `json:"createdAt"`
	UpdatedAt  *time.Time `json:"updatedAt"`
}

File: apps/api/zorde-erp-laboratorio/src/modules/fornecedor/application/dtos/fornecedor.dto.ts
import { z } from 'zod';
import { StatusPessoa } from '../../../../shared/enums/status-pessoa.enum';
import { TipoPessoa } from '../../../../shared/enums/tipo-pessoa.enum';

export const updateFornecedorSchema = z.object({
	id: z.coerce.number(),
	nome: z.string().min(3, 'O nome deve ter pelo menos 3 caracteres').optional(),
	email: z.email('E-mail inválido').optional(),
	contato: z.string().optional(),
	tipoPessoa: z.enum(TipoPessoa).optional(),
	documento: z.string().optional(),
	status: z.enum(StatusPessoa).optional(),
	cep: z.string().optional(),
	uf: z.string().optional(),
	cidade: z.string().optional(),
	logradouro: z.string().optional(),
	numero: z.string().optional(),
	bairro: z.string().optional(),
	observacao: z.string().optional(),
	numeroEndereco: z.string().optional(),
	complemento: z.string().optional(),
	ibge: z.string().optional(),
	razaoSocial: z.string().optional(),
	nomeFantasia: z.string().optional(),
});

export const createFornecedorSchema = z
	.object({
		nome: z.string().min(3, 'O nome deve ter pelo menos 3 caracteres'),
		email: z.email('E-mail inválido'),
		contato: z.string().optional(),
		tipoPessoa: z.enum(TipoPessoa, {
			message: 'Tipo de pessoa inválido (FISICA ou JURIDICA)',
		}),
		documento: z.string().min(11, 'Documento inválido'),
		status: z.enum(StatusPessoa).default(StatusPessoa.ATIVO),
		razaoSocial: z.string().optional(),
		nomeFantasia: z.string().optional(),
		cep: z.string().optional(),
		uf: z.string().optional(),
		cidade: z.string().optional(),
		logradouro: z.string().optional(),
		numero: z.string().optional(),
		complemento: z.string().optional(),
		bairro: z.string().optional(),
		ibge: z.string().optional(),
		observacao: z.string().optional(),
		numeroEndereco: z.string().optional(),
	})
	.refine(
		(data) => {
			const clean = data.documento.replace(/\D/g, '');
			if (data.tipoPessoa === TipoPessoa.FISICA) {
				return clean.length === 11;
			} else {
				return clean.length === 14;
			}
		},
		{
			message: 'CPF deve ter 11 dígitos e CNPJ deve ter 14 dígitos',
			path: ['documento'],
		},
	);

export type CreateFornecedorDto = z.infer<typeof createFornecedorSchema>;
export type UpdateFornecedorDto = z.infer<typeof updateFornecedorSchema>;
